Fix issue 2703 (On-the-fly transformation for principal component projection)

rishu235 opened the pull request:
Fixes #2703 

Changes made in this Pull Request:
 - Added a function for inverse-PCA transform in pca.py: PCA.project_single_frame
 - The function takes an integer/list of component(s) onto which the structure is projected
 - The function returns a transformation function which can be applied to a TimeStep object
 - Added a test function which runs PCA.project_single_frame for all PCA components


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

Commit:
Added a function for inverse-PCA transform along selected PC components.

Commit:
Fixed pepcode warnings in pca.py

Added a test function to run the added functions in pca.py

pep8speaks commented:
Hello @rishu235! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2022-04-11 23:08:24 UTC

Commit:
Fixed pep8 issues in the test_pca.py file

rishu235 committed:
Fixed the docstring

Removed 'workflow' variable. 'func' is now passed directly to 'add_transformations'

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,15 @@ def test_transform_universe():
+    pca_run = PSF_pca.run()
+    func = PSF_pca.project_single_frame()
+    u.trajectory.add_transformations(func)
+    u.trajectory[-1]
```
Test actual numbers, not just that the code doesn't crash.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,42 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+        def wrapped(ts):
+            xyz = self._atoms.positions.ravel() - self._xmean
+            dot = np.dot(xyz, self._p_components[:, components])
```
use a more expressive variable name than `dot` (which could also be confused with the numpy function)

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,42 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            components = range(self.results.p_components.shape[1])
+
+        def wrapped(ts):
+            xyz = self._atoms.positions.ravel() - self._xmean
+            dot = np.dot(xyz, self._p_components[:, components])
+            self._atoms.positions = np.reshape(
+                (np.dot(dot, self._p_components[:, components].T)
+                 + self._xmean), (-1, 3)
```
Possibly do everything in one line (better performance?) with [np.linalg.multi_dot](https://numpy.org/doc/stable/reference/generated/numpy.linalg.multi_dot.html).

Also, if we can avoid transposes somehow then that will also speed up — perhaps using [np.einsum](https://numpy.org/doc/stable/reference/generated/numpy.einsum).

rishu235 replied:
> Possibly do everything in one line (better performance?) with [np.linalg.multi_dot](https://numpy.org/doc/stable/reference/generated/numpy.linalg.multi_dot.html).

The latest commit does everything in one line. However, there are two issues with [np.linalg.multi_dot](https://numpy.org/doc/stable/reference/generated/numpy.linalg.multi_dot.html). 

1) When the trajectory is projected onto a single component, the multi_dot method will fail since it requires the PC component matrix to be 2-D.
2) multi_dot is in fact slower because it tries to figure out the order of dot products by calculating the cost for matrix multiplication. np.dot(np.dot(A, B), C) already computes the optimized cost because, in our case, A and C are  1-D arguments. 
> Also, if we can avoid transposes somehow then that will also speed up — perhaps using [np.einsum](https://numpy.org/doc/stable/reference/generated/numpy.einsum).

I tested einsum and saw no significant difference in performance. I continue to use self.transpose method. I looked at the inverse-PCA implementation in some other standard codes and they all seem to use the transpose, either in the PCA transformation or in the inverse-PCA transformation, depending on the conventions for matrix axes.

orbeckst replied:
Thanks for considering everything, I appreciate your thorough and well-reasoned reply!

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,42 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        """Runs an inverse-PCA and generates
+           a trajectory that is projected on selected PCs
+
+        Run PCA class before using this function.
+        To project the trajectory onto the first principal component,
+            pca = PCA(universe, select='backbone').run()
+            func = pca.project_single_frame(components=0)
+
+        To apply the projection on-the-fly, may add as a transformation
+            u.trajectory.add_transformations(func)
```
The synopsis should describe in words what the method does.

Then add an 
```resT
Example
---------
```
section after the *Returns* section.

See the general structure of docs https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard

orbeckst replied:
Use reST markup for code.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,42 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+        Returns
+        -------
+        Returns the projecting function which transforms a timestep
```
At the end, add two empty lines and then
```reST
 
 
.. versionadded:: 2.2.0
```

orbeckst reviewed (changes requested):
Looks like a very good start. There are some conceptual issues to be resolved https://github.com/MDAnalysis/mdanalysis/issues/2703#issuecomment-1086948647 namely
- How to deal with the situation when the number of PCA atoms is not the number of atoms in the trajectory ("extrapolation"/"applying the PCA filter to whole residues")?
- What error/sanity checks to add?

In the meantime I have a number of smaller comments. 

Please also check that the docs look the way that you expect them to by looking at https://mdanalysis--3596.org.readthedocs.build/en/3596/

orbeckst commented:
Eventually we'll also ask you to add yourself to AUTHORS and add an entry for the CHANGELOG.

orbeckst reviewed (changes requested):
To make progress on the PR, please do the following
* address my comments from my previous review
* **focus on only two approaches for applying the transformation**
  1. all atoms are in the PCA mode so the displacement vector can be applied to all atoms
  2. transform _all_ atoms inside a residue that has exactly one displacement vector per residue (e.g. from CA) by the same vector
* raise an exception for all cases that aren't (i) or (ii).

This will produce a useful contribution but will leave open the door for more refined approaches.

Commit:
Tries to fix the Example format in the docstring

Commit:
Tries again to fix Examples in the docstring

rishu235 commented:
@orbeckst, thank you for your comments. I believe, I have addressed your initial comments, except for the Examples in the docstring and CHANGELOG file.

I will implement the class ii approach, along with exception error and the two unresolved comments mentioned above.

Is there a way I can see what the docstring is going to look like on MDAnalysis webpage, before I push commits?

orbeckst commented:
You can build the docs yourself with
```
python setup.py build_sphinx
```
as described in the [User Guide: building documentation](https://userguide.mdanalysis.org/stable/contributing_code.html#building-code-documentation).

I often just push and the let ReadTheDocs build the docs as part of the PR checks, namely the one called **docs/readthedocs.org:mdanalysis** builds and shows the docs at https://mdanalysis--3596.org.readthedocs.build/en/3596/ (the link under [Details](https://mdanalysis--3596.org.readthedocs.build/en/3596/) for the check).

orbeckst commented:
@rishu235 please ping me (i.e., a comment with `@orbeckst`) when you need me to review again. (I am quite busy but I'll try to review within a day.)

Commit:
Adds a feature to extrapolate inverse-PCA

- PCA.project_single_frame takes an anchor atom per residue to extrapolate the PC projection to non-PCA atoms
- raises a ValueError if the numer of displacement vector per residue is not exactly one
- adds comprehensive positive/negative tests in test_pca.py

Commit:
Removes extra colon symbols at the end of Examples

Commit:
Changes a ValueError message

Commit:
Fixes colon symbols at the end of Examples

rishu235 commented:
@orbeckst please review the changes.

The code can now extrapolate the projection, onto selected PCs, to non-PCA atoms. The extrapolation is turned off by default, to make the function faster. It also throws an error to the user, if the number of displacement vector is not exactly one per residue.

I introduced a line break into the existing line of authors in CHANGELOG to keep the line width consistent. I hope that's doesn't go against the rules.

orbeckst commented:
I'll review in a day. In the meantime please resolve conflicts relative to develop. Breaking the contributors line in CHANGELOG is good --- thanks! (When you resolve the merge conflict, add your name at the end and break the line so thta your under the max character per line (79?) that's stated at top of CHANGELOG).

I also saw selections for resid. These are expensive. Try to do it without them.

Commit:
Addresses PR reviewer's initial comments

- adds more test functions to test the return values of pca.py
- removes a varibale with confusing name
- combines two dot prodcuts into one command
- adds version number in the docstring
- adds an Example section in the docstring

Commit:
Adds a ValueError and updates CHANGELOG

- ValueError is thrown when extrapolation is requested with no anchor atom

- CHANGELOG modifies the author list and lists the enhancement

rishu235 commented:
@orbeckst To apply the transformation per residue, don't we need a residue selection?

(Fixed) Please ignore the changes shown in overview.rst file. Those changes are from another merged PR by someone else. I may have messed up in rebasing/resolving. I'll try to fix that tomorrow.

Commit:
Modifies the documentation

- rebases the repo

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,118 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        Returns
+        -------
+        function
+            The projecting function which transforms a timestep.
```
Document the signature of the returned function here.

orbeckst replied:
I wasn't clear here, I think: The return value *is* a function, so your original docs were correct., What I meant was to add in the text ("The projecting function ...") to write something like: The resulting function `f` takes as input a `Timestep` ts, `f(ts)` and returns the Timestep. 

Or something like that. It should be clear that it can be used as a transformation.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,118 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        .. versionadded:: 2.2.0
+        """
+        if not self._calculated:
+            raise ValueError('Call run() on the PCA before projecting')
```
I think that should be `RuntimeError` — unless we have done things differently elsewhere.

rishu235 replied:
`pca.py` has five more instances of this error, all called `ValueError`

orbeckst replied:
Ok, then we stay consistent and keep it that way.

Btw, please let reviewers resolve conversations, at least that's our custom in MDAnalysis. It helps reviewers to keep track of points that they raised.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,118 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+            for i in group.residues.resnums:
+                n_anchor = self._atoms.select_atoms(
+                           f'resnum {i} and {anchor}').n_atoms
```
Doing a selection for every frame in a trajectory is quite slow. Given that this is a method of a class, you have the opportunity to save state. I would cache the selection. Perhaps we _can_ leave performance improvements for later. 

However, there should be a faster way to get your answer than iterating over residues. Something along the line of 
```python
anchors = group.select_atoms(anchor)
group.residues == anchors.residues 
```
should indicate that you cover all residues. To check that you have exactly one anchor per residue, you could use `anchors.groupby("resindices")` (see [groupby()](https://docs.mdanalysis.org/2.1.0/documentation_pages/core/groups.html#MDAnalysis.core.groups.AtomGroup.groupby)) but that still requires loops. Perhaps it is good enough to just check
```python
np.all(anchors.resindices == group.resindices)
```

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,118 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+                    non_pca.select_atoms(f'resnum {i}').positions -= (
+                        self._atoms.select_atoms(f'resnum {i} and {anchor}')
+                            .positions
+                    )
```
Avoid selections in inner loops!!

Instead of using selections to get residues, directly use the groups, possibly using group set operations. 
```python
non_pca_with_anchor = (non_pca.residues - anchor.residues).atoms
```
Write the operationas a single numpy operation  instead of per-residue loop , perhaps along the lines of
```python
non_pca_with_anchor.positions -= anchor.positions
```
... or something along the line.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,118 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+                    non_pca.select_atoms(f'resnum {i}').positions += (
+                        self._atoms.select_atoms(f'resnum {i} and {anchor}')
+                            .positions
+                    )
```
Avoid selections and loops, see above.

orbeckst reviewed (changes requested):
My primary concern is the use of selections and loops over residues which will lead to very poor performance. I added some ideas how you might be able to get around it. Ultimately it might also be good to cache the selections such as the `anchor` atom group and the `non_pca` group but let's worry about the obvious things first.

orbeckst commented:
> To apply the transformation per residue, don't we need a residue selection?

I think it only makes sense to apply the transformations to the same residues that the pca atoms are in. I think your code is already doing that. It just has to raise an error if it cannot do it and in my review I probably did not consider all cases where this may happen. That's something to think about and possibly address in another round of changes.

orbeckst commented:
By the way, I do appreciate that you're tackling a challenging issue. I know that it might feel that it goes terribly slowly. We are concerned with keeping standards in MDAnalysis high so the process takes as long as it has to. I would encourage you to start with your application in parallel.

Commit:
Avoids expensive loops and atomselections

* Replaces expensive recurring operations by one-time operations

* Calculates an extrapolation matrix, before the first transformation

* Catches more errors

* Adds new and slightly modifies existing functions in test_pca.py

rishu235 commented:
@orbeckst I have managed to avoid loops and selections in the wrapped function. There still are atomselections and one `for` loop _outside_ the wrapped function. Since this part is run only once per function call (as opposed to every frame), the code is much faster now.

Thanks for pointing me in the right direction. I was earlier looping through residues in each frame, since each residue has a unique translation vector. The displacement vectors of `anchor` atoms can be grouped together. But, it was not trivial to apply it to a group of non-PCA atoms, since each displacement vector is used for varying number of atoms in each residue.

I now use the fact that atomgroups remain the same through the frames. I calculate an extrapolation matrix, comprising of zeros and ones, which converts the `anchor` atoms' displacment vector (shape[0] = number of residues in `group`) into a displacement vector for the non-PCA atoms (shape[0] = number of atoms in `group`).

Thank you the encouraging words. I am indeed finding it difficult, with the end-of-term duties, to go back to the GSoC proposal while the PR is open. Every kind gesture goes a long way. At the same time, it is good to see a demand for high standard. It gives me more confidence as an MDAnalysis user.

Commit:
Adds minor modifications to the documentation

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            components = range(self.results.p_components.shape[1])
+
+        def wrapped(ts):
+
```
remove empty line or add a one-line doc string

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            non_pca = group - self._atoms
+            pca_res_indices, pca_res_counts = np.unique(
+                self._atoms.resindices, return_counts=True)
+
```
Add a comment that explains what problem you're solving here and what the key idea is. This will be important if someone else wants to debug this code (or learn from it). They need to know your intentions.

rishu235 replied:
added comments to explain the purpose of matrix_extrapolate below this line

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            matrix_extrapolate = np.zeros((non_pca.n_atoms, anchors.n_atoms))
+            cumul_atoms_old = 0
+            for res in group.residues:
+                res_index = res.resindex
```
I'd just use res.resindex to keep the code more concise, the time saving from caching the attribute access are not worth it, I think.

rishu235 replied:
done

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+            if group is not None:
+                anchors_pos = np.matmul(matrix_extrapolate, anchors.positions)
+                non_pca.positions -= anchors_pos
```
Make it one calculation.

orbeckst replied:
An you can just use `@` which is the recommended operator.

orbeckst replied:
Actually, do _not_ make it one calculation but keep `anchors_pos` because you can reuse them for the backtransformation below! (At least if I understood the code correctly.)

rishu235 replied:
added the `@` operator.

> ... keep `anchors_pos` because you can reuse them for the backtransformation below! (At least if I understood the code correctly.)

`anchors` are (required to be) part of `self` atomgroup, which is projected with the inverse-PCA in the next line. So, need to recalculate later.

orbeckst replied:
Thanks for the explanation, I missed the fact that anchors are also being transformed as part of atomgroup.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        def wrapped(ts):
+
+            if group is not None:
+                anchors_pos = np.matmul(matrix_extrapolate, anchors.positions)
```
This is a clever solution to partition the translation vector over residues, basically a broadcast with uneven partitions.

Just out of curiosity, does numpy have a function for that @tylerjereddy ?

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            )
+
+            if group is not None:
+                anchors_pos = np.matmul(matrix_extrapolate, anchors.positions)
```
Why do you have to recalculate anchors_pos? It should be the same as before, so you can just use the variable that you defined previously.

(And then there's an excellent reason to _not_ put everything on one line... I'll edit my comment)

orbeckst replied:
I was wrong, thanks to the explanation above. Needs to be recalculated.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+            xyz = self._atoms.positions.ravel() - self._xmean
+            self._atoms.positions = np.reshape(
+                (np.dot(np.dot(xyz, self._p_components[:, components]),
```
I _think_ you can use @ instead of dot(). Check the docs.

rishu235 replied:
The first operation results in a _scalar_, when only one PC is selected, and gives an error with @

orbeckst replied:
Well, if the current code is robust then we go with this code! Thanks for checking.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            if group is not None:
+                anchors_pos = np.matmul(matrix_extrapolate, anchors.positions)
+                non_pca.positions += anchors_pos
+
```
remove blank

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,129 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+                non_pca.positions += anchors_pos
+
+            return ts
+
```
remove blank

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
 
+def test_project_no_pca_run(u, pca):
+    pca_class = PCA(u, select=SELECTION)
+    with pytest.raises(ValueError):
```
also check the error message with the kwarg of pytes.raises

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+
+def test_project_none_anchor(u, pca):
+    group = u.select_atoms('resnum 1')
+    with pytest.raises(ValueError):
```
also check the error message with the kwarg of pytes.raises

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+
+def test_project_more_anchor(u, pca):
+    group = u.select_atoms('resnum 1')
+    with pytest.raises(ValueError):
```
also check the error message with the kwarg of pytest.raises

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+
+def test_project_less_anchor(u, pca):
+    group = u.select_atoms('all')
+    with pytest.raises(ValueError):
```
also check the error message with the kwarg of pytest.raises

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+def test_project_improper_anchor(u):
+    pca = PCA(u, select='name CA').run()
+    group = u.select_atoms('all')
+    with pytest.raises(ValueError):
```
also check the error message with the kwarg of pytest.raises

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+
+
+def test_project_compare_projections():
+    u = mda.Universe(PSF, DCD)
```
make the universe a fixture
```python
@pytest.fixture
def universe():
   return mda.Universe(PSF, DCD)

def test_xxxxx(universe):
    u_copy = universe.copy()
    ...
```

Same for the other ones below.

rishu235 replied:
Done. It's called u_fresh; gives a fresh universe to each function.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+    u.trajectory[0]
+    coord1 = project1(u.trajectory.ts).positions.copy()
+
+    with pytest.raises(AssertionError):
```
delete the line, 
```suggestion
```
just use 
```python
assert_almost_equal(coord0, coord1, decimal=5)
```
You don't have to wrap the assert_* functions.

orbeckst replied:
Or do you expect this situation to fail? Then mark the test xfail. 

But please explain what this test is supposed to show.

rishu235 replied:
verifies that the projections along two different PCs are different. Added "assert not np.allclose()" because we want it to detect future changes in the code, in case it returns the same timestep (or the same projection).

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+
+    if case == 1:
+        assert_almost_equal(distances_original, distances_new, decimal=5)
+    if case == 2:
```
make it `elif`

rishu235 replied:
no case numbers in the new version. changed second if to else.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+        assert_almost_equal(distances_original, distances_new, decimal=5)
+    if case == 2:
+        with pytest.raises(AssertionError):
+            assert_almost_equal(distances_original, distances_new, decimal=5)
```
add an `else` that raises NotImplementedError to catch any test misconfigurations

rishu235 replied:
replaced this if by else

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+                                       anchor='resnum 1 and name CA')
+
+    distances_original = (
+        mda.lib.distances.self_distance_array(group.positions.copy())
```
I think you can remove the copy() here and below as the data are not altered and a new data structure is returned.

rishu235 replied:
Done. All the copy() are removed, except where the function is applied twice in succession.

orbeckst replied:
Furthermore, `ATomGroup.positions` is already a copy!

rishu235 replied:
Is it? This may require a further investigation of the `positions` method. I have a `.copy()` method in test_project_twice_projection (line 230) and I absolutely need it. ~~(Copy in line 231 can be removed though. I'm on it.)~~

I also compared the actual numbers. The test works as intended with`project_first(u_fresh.trajectory.ts).positions.copy()`, but not with `project_first(u_fresh.trajectory.ts).positions`

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +145,101 @@ def test_transform_universe():
+    if case == 1:
+        assert_almost_equal(distances_original, distances_new, decimal=5)
+    if case == 2:
+        with pytest.raises(AssertionError):
```
```suggestion
```

orbeckst replied:
Or do you want to say that you expect case 2 to fail? If so, then you should use xfail marks.

rishu235 replied:
Removed parametrization of case number in this test, since expected bugs will be caught by other tests (twice_projection and compare_projection)

orbeckst reviewed (changes requested):
Neat idea with the matrix multiplication to generate the correct shape of the trasnlation vector array.

Please have a look at my comments. Thanks.

orbeckst commented:
One comment on how we do reviews in MDA: We leave comments open until the _reviewer_ closes them. You as someone addressing comments would add a comment ("done" or an explanation or sometimes nothing and just fix the code) and the reviewer would close them when they are satisfied. This makes reviewing A LOT easier. We maybe reviewing 4 PRs at the same time and it's sometimes difficult to keep track of everything. Everything that makes your reviewers' life easier is good and will help them to spend time on your PR.

Don't worry, part of the process is to learn the customs. And you're doing an excellent job at discussing the PR with me.

orbeckst commented:
@rishu235 how big does the extrapolation matrix become and how much of a computational bottleneck will it become if you apply it to a protein with, say 20,000 atoms? Can you do a quick `%timeit` of numpy matrix multiplication with random matrices and vectors for a few sizes? 

I like your approach conceptually and accept it as a solution to the problem. If the performance becomes a problem for larger systems then we will open a new issue for performance improvement.

orbeckst commented:
I think the PR is almost done. Once you address my remaining comments I will have one last request, and I'll tell you right away:

Could you demonstrate your approach by making a movie of the test trajectory (PSF, DCD) when projected on the first PC? Add the code for how you do it to the documentation (eventually it can be moved to the User Guide in the [PCA](https://userguide.mdanalysis.org/stable/examples/analysis/reduced_dimensions/pca.html) section). 

If you're adventurous, you can also try it for https://www.mdanalysis.org/MDAnalysisData/adk_equilibrium.html .

Commit:
Cleans up the functions in pca.py and test_pca.py

- replaces numpy.matmul operators are by @

- pytest.raises checks for error messages

- xfail mark added for the negative tests

- adds comments in the pca.py and test_pca.py

Commit:
Removes an obsolete print line

rishu235 commented:
> how big does the extrapolation matrix become and how much of a computational bottleneck will it become if you apply it to a protein with, say 20,000 atoms?

It does become quite expensive relative to other parts. The matrix has the size of (`n_atoms`,`n_residues`). For a thousand residues with twenty atoms, one matrix multiplication took 261 ms on my computer. This is compared to 1.9 ms for projecting the PCA atoms (equivalent random matrix operations rather). ~~I should at least avoid one of the two matrix multiplications in the wrapped function~~ (Pushed changes). For 100 residues, each with twenty atoms, one matrix multiplication takes 1.58 ms, while the PCA atoms took 205 µs.


> Could you demonstrate your approach by making a movie of the test trajectory (PSF, DCD) when projected on the first PC? Add the code for how you do it to the documentation (eventually it can be moved to the User Guide in the [PCA](https://userguide.mdanalysis.org/stable/examples/analysis/reduced_dimensions/pca.html) section).

Yes, I was already thinking of that. I also have a trajectory filtered by aligning each residue. Yes, it's very expensive, but it looks super cool! @orbeckst Could I post the three movies (raw, anchor_filtered, and align_filtered) on the issue #2703  page, instead of here, and that will serve the next person working on the issue?

rishu235 commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,128 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        def wrapped(ts):
+            """Projects a timestep"""
+            if group is not None:
+                non_pca.positions -= matrix_extrapolate @ anchors.positions
```
Note to self: Avoid this matrix multiplication. Save the old anchor coordinates to calculates the displacement vector.
Finally,

```
non_pca.positions += matrix_extrapolate @ (anchors.positions_old - anchor.positions_new)
```

Commit:
Avoids an expensive matrix multiplication

- wrapped() saves the old anchor coordinates to get the displacement vector

- links `Timestep` in the dpcumentation to `Timestep` definition page

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+            project = pca.project_single_frame(components=0)
+
+        To projet onto the first two components, run::
```
```suggestion
        To project onto the first two components, run::
```

rishu235 replied:
Done

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            The resulting function f(ts) takes as input a
+            :class:`~MDAnalysis.coordinates.base.Timestep` ts,
+            and returns ts with the projected structure
+
```
I think we need to include a warning:
```
.. warning::

   The transformation function takes a :class:`Timestep` as input because this is required for using
   it as a :ref:`transformations`. However, the transformations are all applied on the group and atoms
   of the Universe that was used for the PCA. It is *expected* that the `ts` is from the same Universe
   but this is currently not checked.
```

rishu235 replied:
added below

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        anchor : string, optional
+            The string to select the PCA atom whose displacement vector
+            is applied to non-PCA atoms in a residue. The atomselection
+            must have exactly one PCA atom in each residue of ``group``.
```
Insert:

"The `anchor` selection is applied to `group`."

rishu235 replied:
Added

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+            # matrix_extrapolate is later multiplied to anchors' coordinates.
+            # multiplication shapes the array as appropriate for non_pca atoms
+            matrix_extrapolate = np.zeros((non_pca.n_atoms, anchors.n_atoms))
```
Set the dtype to float32 or int — I don't know which one will be faster. Our coordinates are all float32 so we don't want float64 (default) unless we need it.

Or, if you really need float64 then be explicit about it.

rishu235 replied:
Set the dtype to np.int8, the performance is comparable to the default allocation. int may improve memory efficiency though.
I tested a few types on my machine, and the default float64 is as fast as/faster than float32.

orbeckst replied:
You can even make the matrix Boolean, which makes the matrix multiplication a bit slower compared to int8, at least when I played with 100x100 matrices. 

For me, the fastest appears to be everything float32 or float64:
```
bool @ f64: 14.9 µs ± 278 ns
bool @ f32: 13.8 µs ± 735 ns
f64 @ f64 : 3.4 µs ± 143 ns
f32 @ f64:  11 µs ± 232 ns
f32 @ f32 : 3.56 µs ± 269 ns
i8 @ f32 : 11.1 µs ± 411 ns
i8 @ f64 : 13.3 µs ± 2.23 µs
```
Given that speed is an issue, I would say we do not care about memory but ensure that **there is no implicit type casting** involved: Ensure that **all terms in the matrix multiplication have the same type**.

orbeckst replied:
```python
import numpy as np
rng = np.random.default_rng()
Mbool = rng.random((100, 100)) > 0.5
Mf64 = Mbool.astype(np.float64)
Mf32 = Mbool.astype(np.float32)
vf64 = rng.random(100)
vf32 = v.astype(np.float32)

%timeit Mbool @ vf64
%timeit Mf32 @ vf32
...
```

rishu235 replied:
I see! I was playing with the dtype of only one matrix. No wonder, float32 was slower than float64, since the second matrix was implicitly defined as float64. Thank you!

Displacement vector is float32, so I've made `matrix_extrapolate` float32

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        or apply the projection to the universe::
+
+            u.trajectory.add_transformations(project)
+
```
Can you add a section "Background" where you essentially write out the equations for the transformation?

Also write out the mathematics of your extrapolation trick.

rishu235 replied:
Added before the Examples section, going by the section order at style guide at https://numpydoc.readthedocs.io/en/latest/format.html#notes

Commit:
Fixes a typo

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+                cumul_atoms_old = cumul_atoms_new
+
+        if components is None:
+            components = range(self.results.p_components.shape[1])
```
Using a `range` object works for indexing a np.array but is ~50x slower for me. I'd do
```suggestion
            components = np.arange(self.results.p_components.shape[1])
```
and store the array.

```python
In [1]: import numpy as np

In [2]: a = np.arange(100)

In [3]: a[range(4)]
Out[3]: array([0, 1, 2, 3])

In [4]: a[range(29)]
Out[4]:
array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28])

In [5]: %timeit a[range(29)]
9.29 µs ± 61.7 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: ix = np.arange(29)

In [7]: %timeit a[ix]
209 ns ± 7.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [8]: 9.29e-6/209e-9
Out[8]: 44.449760765550245
```

rishu235 replied:
Fixed. I think, it was a typo! An oversight that _seemed_ to work.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+        def wrapped(ts):
+            """Projects a timestep"""
+            if group is not None:
+                anchors_coords_old = anchors.positions
```
(For a moment I thought you need a copy() here but then I remembered that AtomGroup.positions is *always* a new array because we use fancy indexing `Timestep.positions[ag.ix]`.)

All good!

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+            )
+
+            if group is not None:
+                non_pca.positions += (matrix_extrapolate @
+                                      (anchors.positions - anchors_coords_old))
```
Smart! 

I look forward to reading the equations in the Background section.

I have a feeling that we could get the rearrangement more cheaply but I still like the very clean formulation as a matrix multiplication.

rishu235 replied:
Added equations to show PCA transformation and inverse-PCA transformation. Explained the extrapolation trick too.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -41,6 +41,12 @@ def u():
     return mda.Universe(PSF, DCD)
 
 
+@pytest.fixture(scope='module')
```
I think this should be function scoped because the projection will change the underlying universe. Thus the tests cannot be safely run in parallel. Creating the universe is not terribly expensive so rather be safe (make a new one for every test) than sorry (trying to be fast but sometimes wrong).

rishu235 replied:
Done.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +151,113 @@ def test_transform_universe():
+    assert "Some 'anchors' are not part of PCA class" in str(exc.value)
+
+
+@pytest.mark.xfail
```
Now I understand the test a bit better. You expect this to be different. Then you do _not_ use xfail (which we use when we know that something hasn't been fixed).

Instead, invert the assertion below.

orbeckst replied:
How important is this test, i.e., when do you think it would fail to pass, i.e., what would have gone wrong if `coord1 == coord2`?

rishu235 replied:
Will catch if the projection does nothing, or is insensitive to the `components` variable.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +151,113 @@ def test_transform_universe():
+    coord0 = project0(u_project.trajectory.ts).positions
+    u_project.trajectory[0]
+    coord1 = project1(u_project.trajectory.ts).positions
+    assert_almost_equal(coord0, coord1, decimal=5)
```
Need to rewrite
```python
assert not np.allclose(coord0, coord1)
```

rishu235 replied:
Done

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +151,113 @@ def test_transform_universe():
+
+    coord_original = u.trajectory.ts.positions
+    coord_reconstructed = project(u_project.trajectory.ts).positions
+    assert_almost_equal(coord_original, coord_reconstructed, decimal=5)
```
recommended (from https://numpy.org/doc/stable/reference/routines.testing.html )
```suggestion
    assert_allclose(coord_original, coord_reconstructed, rtol=1e-5)
```
We still have mostly assert_almost_equal but eventually we are going to replace them.

rishu235 replied:
Done

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +151,113 @@ def test_transform_universe():
+    u_project.trajectory[0]
+    coord1 = project_first(u_project.trajectory.ts).positions.copy()
+    coord2 = project_second(u_project.trajectory.ts).positions.copy()
+    assert_almost_equal(coord1, coord2, decimal=5)
```
use assert_allclose

rishu235 replied:
Done

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +151,113 @@ def test_transform_universe():
+@pytest.mark.parametrize(
+    ("n1", "n2"),
+    [(0, 0), (1, 1), ([0, 1], [0, 1]),
+     pytest.param(0, 1, marks=pytest.mark.xfail),
+     pytest.param(1, 0, marks=pytest.mark.xfail)]
```
These two are always expected to fail. But if for some reason they pass, the tests will just show "XPASS" and not fail so these tests are not alerting us to any errors that might be introduced in the code. Just remove the two cases. If you want to test them because you're concerned that something could happen to the code where suddenly the two frame projections are the same, write a new test.

orbeckst replied:
(You did use marks really nicely!)

rishu235 replied:
Thank you :)

And now I've removed them all :/
because we don't want XPASS

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +151,113 @@ def test_transform_universe():
+@pytest.mark.parametrize(
+    "sel",
+    ['resnum 1 and name CA CB CG',
+     pytest.param('resnum 1 and name N CB CG', marks=pytest.mark.xfail)]
```
Is this something that should eventually work? Then leave it here with xfail.

However, this is one of the cases that you catch with a ValueError. Therefore, add a separate test `test_project_ValueError_toomany_anchors` or something like that. Importantly, the test should be testing (and making it clear) that it wants to see an exception because you're constructing the failure case.

rishu235 replied:
No, this should always fail.
The purpose is to show that anchor atom (and only the anchor atom) indeed works like an anchor. Removed the second test case though since the errors will be caught by other tests anyway.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_pca.py:
```diff
@@ -145,6 +151,113 @@ def test_transform_universe():
+    ['resnum 1 and name CA CB CG',
+     pytest.param('resnum 1 and name N CB CG', marks=pytest.mark.xfail)]
+)
+def test_projet_extrapolate_translation(u_project, sel):
```
project --> project

rishu235 replied:
Done

orbeckst reviewed (changes requested):
Please see comments, sorry didn't have as much time to review as I wanted.

Commit:
Modifies the test functions

- removes xfail marks

- replaces assert_almost_equal by assert_allclose

- adds a warning in the documentations

Commit:
Adds a notes section

- explains the background of inverse-PCA

- explains the extrapolation trick

Commit:
Fixes indentation in docstring

Commit:
Introduces a line break

- The equations in Notes were earlier appearing on the same line

orbeckst reviewed (changes requested):
This looks really good, ❤️ it!

The only remaining discussion point is the speed of the matrix multiplication. I played around as detailed in the comment and it seems to me that we should avoid any kind of implicit type casting here. Could you please write the code so that the type of the extrapolation matrix is the same type as the displacement vector? I don't know what the type of the displacement vector is — I *assumed* float32 but maybe some upcasting to float64 happens somewhere. 

Prioritize speed over memory efficiency.

Commit:
Changes large matrices' type

- matrix_extrapolate is defined as np.float32, consistent with .positions

rishu235 commented:
@orbeckst Looping through 97 frames in a protein trajectory, testfiles (PSF, DCD) with 214 residues and 3341 atoms, takes

1) 13.3 ms ± 280 µs with no transformation
raw trajectory

2) 26.8 ms ± 3.44 ms with `project(0)`
inverse-PCA for the PCA atoms (backbone) only and non-PCA atoms untouched

3) 66.3 ms ± 8.21 ms with `project(0, group=all, anchor='name CA')`
inverse-PCA for the PCA atoms and extended to non-PCA atoms by anchoring to a PCA atom in each residue.

4) Not implemented yet
inverse-PCA for the PCA atoms and extended to non-PCA atoms by aligning the PCA-atoms in each residue.

I have made ten-second movies. You may have to download them for a good resolution. Please watch the [movies](https://uofc-my.sharepoint.com/:f:/g/personal/rishabh_shukla_ucalgary_ca/EpBePX_TWP9LpNjt4XwoCeIBkqfptRKANXmOMjscODyPpQ?e=yfWYy1) in the order (and after the fourth movie, remember to pick up your jaw from the floor!)

orbeckst commented:
The movies look very cool. Can you describe what the differences between them are? From the file names I only understand "raw".

EDIT: Reading your post again makes things clearer. I was confused by "inverse" -- for me this is just "projection on PC".   Is 4 taking the initial frame for the "non PCA atoms" to align?

orbeckst reviewed (changes requested):
I have a number of points to be addressed, after thinking things through and reading your docs in detail

# speeding up the extrapolation
I think I have an idea how to avoid the matrix multiplication, which should be much faster. Your matrix contains mostly zeros and what it really does is describe a way to *fancy index* the displacement vector `v`: Let's say you have
```python
>>> v = np.arange(4) * 10
array[0, 10, 20, 30]
```
and you want to make it 
```python
array([ 0,  0, 10, 10, 10, 20, 20, 30, 30, 30, 30, 30])
```
then you can either set up your matrix that starts as (EDIT: I incorrectly wrote the transpose of the matrix ... but you know what I mean)
```
1 0 0 0 0 0 0 ...
1 0 0 0 0 0 0 ...
0 1 0 0 0 0 0 ...
0 1 0 0 0 0 0 ...
0 1 0 0 0 0 0 ...
0 0 1 0 0 0 0 ...
...
```
and matrix multiply or you can fancy index:
```
v_expanded = v[[0, 0, 1, 1, 1, 2, 2, 3, 3, 3, 3, 3]]
```

Generating the fancy-index vector should be easy as you've essentially done it already.

Please test this approach and get some timings. I am pretty sure its faster and it definitely conserves memory.

# doc consistency
The User Guide's [PCA tutorial](https://userguide.mdanalysis.org/stable/examples/analysis/reduced_dimensions/pca.html#Visualising-projections-into-a-reduced-dimensional-space) already lays out the basic equations. In particular, what you call PC_score is called weight w_i there. It also uses **u_i** for PC_i. Could you please change your equations to be consistent with the User Guide. 

Note that the User Guide incorrectly bold-faces w_i(t), implying it is a vector. However, it's a scalar and should be set in standard math mode italics. (The equation for the projection with the outer produce is not correct because *for a given time t*, the multiplication is simply scalar-vector multiplication *w_i(t)* **u_i**. The Python implementation does correctly use `np.outer` because the weights `trans1` are _all_ times. )

orbeckst commented:
For generating the index vector for the expansion look into the various numpy index tricks such as [np.r_](https://numpy.org/doc/stable/reference/generated/numpy.r_.html) and similar index generation objects.

Commit:
Uses fancy indexing for better speed

- Removes matrix multiplication in each frame

- Makes sure that residues are transformed, not random atoms

- Makes the equations consistent with PCA tutorial

- Modifies the explanation of the extrapolation trick

rishu235 commented:
>Is 4 taking the initial frame for the "non PCA atoms" to align?

The fourth approach, not part of this PR (right?), aligns the raw PCA atoms to projected PCA atoms and applies this transformation to the non-PCA atoms. Basically, it accounts for displacement _and rotation_.

Fancy indexing is simpler and faster. The time taken to loop through 97 frames are now
- 13.3 ms ± 99.3 µs for raw
- 21.1 ms ± 749 µs for projected backbone
- 36.4 ms ± 1.04 ms for projected backbone and projection extended by anchoring to CA atoms

The first two are more or less the same, as expected. While, the third one is almost twice as fast.

Also made the equations consistent with [PCA tutorial](https://userguide.mdanalysis.org/stable/examples/analysis/reduced_dimensions/pca.html#Visualising-projections-into-a-reduced-dimensional-space).

Commit:
slighltly modifies the Notes section

Under extrapolation, atoms changed to non-PCA atoms for clarity.

Commit:
Clarifies comments

orbeckst commented:
Great work. I might not get to reviewing over the weekend (work/life balance…) so please have some patience with me. Maybe another coredev has some comments. 

> Am 4/9/22 um 13:24 schrieb rishu235 ***@***.***>:
> 
> ﻿
> Is 4 taking the initial frame for the "non PCA atoms" to align?
> 
> The fourth approach, not part of this PR (right?), aligns the raw PCA atoms to projected PCA atoms and applies this transformation to the non-PCA atoms. Basically, it accounts for displacement and rotation.
> 
> Fancy indexing is simpler and faster. The time taken to loop through 97 frames are now
> 
> 13.3 ms ± 99.3 µs for raw
> 21.1 ms ± 749 µs for projected backbone
> 36.4 ms ± 1.04 ms for projected backbone and projection extended by anchoring to CA atoms
> The first two are more or less the same, as expected. While, the third one is almost twice as fast.
> 
> Also made the equations consistent with PCA tutorial.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

rishu235 commented:
Sure! Thank you for your prompt responses :)

> I might not get to reviewing over the weekend (work/life balance…)

I'm actually glad to hear that! Believe it or not, I had (almost) decided _not_ to apply for GSoC. Now, I'll think about it again.

orbeckst commented:
> I'm actually glad to hear that! Believe it or not, I had (almost) decided not to apply for GSoC. Now, I'll think about it again.

Please consider applying! You have shown that you have the initiative to tackle difficult problems, the creativity to come up with solutions, the humility to listen to advice and criticism, and the professionalism needed to engage in productive discussions that ultimately lead to great code that will be useful to many others.

I have very much enjoyed the discussion on this PR and I am thrilled to see it getting to completion. I have very little time in my day-to-day job so I am trying to make time when I can to review. Therefore, you'll get comments at odd hours, and sometimes the comments are fairly terse because that's all I can squeeze into the gap between two meetings or teaching. Perhaps I should have made clear that we don't expect you to respond immediately, we understand that people have lives. 

I really want to see your excellent work completed — and I would very much look forward reading a GSOC proposal from you. (Feel free to ask questions about the proposal on the developer list. You can also submit the proposal already and embed a link to a draft such as Google Doc or Overleaf.)

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,174 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+            # non_pca has "all" the atoms in residues of `group`. This makes
+            # sure that extrapolation works on residues, not random atoms.
+            non_pca = group.residues.atoms[:] - self._atoms
```
No need for slicing
```suggestion
            non_pca = group.residues.atoms - self._atoms
```

orbeckst reviewed (approved):
Looking excellent, great work.

(You can raise an issue to add the fitting procedure for extrapolation and if you already have code then please open a PR. Reviewing might be a bit slow because we prioritized GSOC at the moment and you already have an excellent PR done.)

orbeckst committed:
Update package/MDAnalysis/analysis/pca.py

orbeckst committed:
Merge branch 'develop' into fix-issue-2703

rishu235 commented:
Thank you for the review and approval. Does this PR need approvals from other coredevs before merging?

> I have very much enjoyed the discussion on this PR and I am thrilled to see it getting to completion. I have very little time in my day-to-day job so I am trying to make time when I can to review. Therefore, you'll get comments at odd hours...

I wasn't sure how my previous comment would be read. But you do get the point. My recent disinterest is mainly on me (more blah blahs to come in the google group). I am hoping I would again be excited towards the application soon.

> I really want to see your excellent work completed — and I would very much look forward reading a GSOC proposal from you. 

I too have been enjoying the review process and I appreciate all the encouragement and support from you ❤️

> (You can raise an issue to add the fitting procedure for extrapolation and if you already have code then please open a PR. Reviewing might be a bit slow because we prioritized GSOC at the moment and you already have an excellent PR done.)

I have the quick-and-dirty code in a notebook, but it will take some effort to make it nice with the appropriate tests and documentation. I will work on that after a couple of weeks, I think.

Meanwhile, I will try to increase the overlap between my GSoC proposal and my PhD project, which is the main constraint with respect to time and priority. If the overlap looks good enough to me and my supervisor, I will run it by someone via the google group later this week.

orbeckst commented:
There's one failure in the 32 bit windows run https://dev.azure.com/mdanalysis/mdanalysis/_build/results?buildId=3458&view=logs&j=8c1244f7-4387-51eb-cf10-2a1290f32f77&t=6cc8a0ec-9d5a-5758-2f06-db7cff0afb7f

There's an implicit int64 -> int32 case (maybe in repeat()) that needs to be made explicit. 

Sorry, I'll have to raise this as a blocking review. Maybe @tylerjereddy or @IAlibay has a good idea.

orbeckst commented on package/MDAnalysis/analysis/pca.py:
```diff
@@ -414,6 +414,174 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+                non_pca_atoms = np.append(non_pca_atoms,
+                                          res.atoms.n_atoms - n_common)
+            # index_extrapolate records the anchor number for each non-PCA atom
+            index_extrapolate = np.repeat(np.arange(anchors.atoms.n_atoms),
```
There's one failure in the 32 bit windows run https://dev.azure.com/mdanalysis/mdanalysis/_build/results?buildId=3458&view=logs&j=8c1244f7-4387-51eb-cf10-2a1290f32f77&t=6cc8a0ec-9d5a-5758-2f06-db7cff0afb7f

There's an implicit int64 -> int32 case (maybe in repeat()) that needs to be made explicit.

rishu235 replied:
Fixed now.

If I understand correctly, it was because I had an array explicitly defined as np.int64, which later required an implicit conversion on a 32-bit machine. I changed the array creation line to int. The new declared array does not require a conversion later.

orbeckst replied:
you solved it – test passes!

orbeckst reviewed (changes requested):
We need to figure out why the 32 bit windows fails.

Commit:
changes dtype from np.int64 to int

Commit:
Merge branch 'fix-issue-2703' of https://github.com/rishu235/mdanalysis into fix-issue-2703

orbeckst commented:
I'll squash-merge the PR in about an hour.

(After getting a coffee ☕ and probably after my next meeting.)

orbeckst commented:
Congratulations @rishu235 , you added a really neat feature to MDAnalysis! 💯 🏅 

I closed #2703 because you solved the key problems but your idea of using structural alignment is really cool so _please_ just open a follow-up issue where you mention #2703 and this PR and then say what you would want to see next. You don't have to open a PR — you can do it if/whenever you want to, or someone else could attempt it and you could review it and give advice. Either way, we use issues as reminders for how to move forward.

Eventually we'd also want to update the [User Guide: PCA](https://userguide.mdanalysis.org/stable/examples/analysis/reduced_dimensions/pca.html) page and include an example in the [User Guide: On-the-fly transformations](https://userguide.mdanalysis.org/stable/trajectories/transformations.html).
